klient/k8s/watcher/watch.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watcher
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+)
+
+// EventHandlerFuncs is an adaptor to let you easily specify as many or
+// as few of functions to invoke while getting notification from watcher
+type EventHandlerFuncs struct {
+	addFunc     func(obj interface{})
+	updateFunc  func(newObj interface{})
+	deleteFunc  func(obj interface{})
+	watcher     watch.Interface
+	ListOptions *cr.ListOptions
+	K8sObject   k8s.ObjectList
+	Cfg         *rest.Config
+	Ctx         context.Context
+}
+
+// EventHandler can handle notifications for events that happen to a resource.
+// Start will be waiting for the events notification which is responsible
+// for invoking the registered user defined functions.
+// Stop used to stop the watcher.
+type EventHandler interface {
+	Start()
+	Stop()
+}
+
+// Start triggers the registered methods based on the event received for
+// particular k8s resources
+func (e *EventHandlerFuncs) Start() error {
+	cl, err := cr.NewWithWatch(e.Cfg, cr.Options{})
+	if err != nil {
+		return err
+	}
+
+	w, err := cl.Watch(e.Ctx, e.K8sObject, e.ListOptions)
+	if err != nil {
+		return err
+	}
+
+	// set watcher object
+	e.watcher = w
+
+	for event := range e.watcher.ResultChan() {
```

Comment:
> If an error occurs or Stop() is called, the implementation will close this channel and release any resources used by the watch.

The `ResultChan` will close if there is an error that occurs right ? Do we need a way to re-register the watcher in those cases? Something similar to `retry.OnError` helper of sort ?

Reply:
If an error occurs, let the user restart the watch instead of attempting to recover with a retry (I think).